File: apps/mobile/plugins/OwndayWidget.swift
import AppIntents
import SwiftUI
import WidgetKit
private struct Habit: Identifiable {
  let id: String; let title: String; var done: Bool; var value: Double; let target: Double?
  var revision: Int; var dependsOn: String?
}
private struct Snapshot {
  let profile: String; let localDate: String; let ru: Bool; var habits: [Habit]
}
private func calendar(_ root: [String: Any]) -> Calendar {
  var calendar = Calendar(identifier: .gregorian)
  calendar.timeZone = TimeZone(identifier: root["timezone"] as? String ?? "UTC") ?? TimeZone(secondsFromGMT: 0)!
  return calendar
}
private func readSnapshot(_ directory: URL, at: Date = Date()) throws -> Snapshot {
  let root = widgetRoot(directory), calendar = calendar(root)
  let logical = calendar.component(.hour, from: at) < (root["dayStartHour"] as? Int ?? 4) ? calendar.date(byAdding: .day, value: -1, to: at)! : at
  let parts = calendar.dateComponents([.year, .month, .day], from: logical)
  let date = String(format: "%04d-%02d-%02d", parts.year!, parts.month!, parts.day!)
  let days = root["days"] as? [[String: Any]] ?? []
  let day = days.first { $0["localDate"] as? String == date }
  var habits = (day?["habits"] as? [[String: Any]] ?? []).compactMap { item -> Habit? in
    guard let id = item["id"] as? String, let title = item["title"] as? String else { return nil }
    return Habit(id: id, title: title, done: item["done"] as? Bool ?? false, value: item["value"] as? Double ?? 0, target: item["target"] as? Double, revision: item["revision"] as? Int ?? 0, dependsOn: item["dependsOn"] as? String)
  }
  let profile = root["profileId"] as? String ?? ""
  let acknowledged = Set(root["acknowledgedOperationIds"] as? [String] ?? [])
  for queued in try widgetPending(directory) {
    guard queued["profileId"] as? String == profile, let change = queued["change"] as? [String: Any], change["localDate"] as? String == date,
      let id = change["operationId"] as? String, !acknowledged.contains(id), let index = habits.firstIndex(where: { $0.id == (change["habitId"] as? String) }), let action = change["action"] as? [String: Any] else { continue }
    habits[index].done = action["kind"] as? String != "clear"
    habits[index].value = habits[index].done ? (action["value"] as? Double ?? habits[index].target ?? 1) : 0
    habits[index].revision += 1
    habits[index].dependsOn = id
  }
  return Snapshot(profile: profile, localDate: date, ru: root["locale"] as? String != "en", habits: habits)
}
struct ToggleHabitIntent: AppIntent {
  static var title: LocalizedStringResource = "Mark habit"
  static var openAppWhenRun = false
  @Parameter(title: "Habit") var habitId: String
  @Parameter(title: "Profile") var profileId: String
  @Parameter(title: "Date") var localDate: String
  init() {}
  init(habitId: String, profileId: String, localDate: String) { self.habitId = habitId; self.profileId = profileId; self.localDate = localDate }
  func perform() async throws -> some IntentResult {
    try withWidgetLock { directory in
      let snapshot = try readSnapshot(directory)
      guard snapshot.profile == profileId, snapshot.localDate == localDate, let habit = snapshot.habits.first(where: { $0.id == habitId }) else { return }
      let id = UUID().uuidString.lowercased()
      let action: [String: Any] = habit.done ? ["kind": "clear"] : ["kind": "set", "status": "done", "value": habit.target ?? 1]
      var change: [String: Any] = ["operationId": id, "habitId": habitId, "localDate": localDate, "baseRevision": habit.revision, "action": action]
      if let dependency = habit.dependsOn { change["dependsOn"] = dependency }
      let formatter = ISO8601DateFormatter(); formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
      let queued: [String: Any] = ["profileId": profileId, "change": change, "createdAt": formatter.string(from: Date())]
      try JSONSerialization.data(withJSONObject: queued).write(to: directory.appendingPathComponent(id + ".json"), options: .atomic)
    }
    WidgetCenter.shared.reloadAllTimelines()
    return .result()
  }
}
private struct Entry: TimelineEntry { let date: Date; let snapshot: Snapshot }
private struct Provider: TimelineProvider {
  func placeholder(in context: Context) -> Entry { Entry(date: Date(), snapshot: Snapshot(profile: "", localDate: "", ru: true, habits: [])) }
  func getSnapshot(in context: Context, completion: @escaping (Entry) -> Void) {
    completion((try? withWidgetLock { Entry(date: Date(), snapshot: try readSnapshot($0)) }) ?? placeholder(in: context))
  }
  func getTimeline(in context: Context, completion: @escaping (Timeline<Entry>) -> Void) {
    let entries = try? withWidgetLock { directory -> [Entry] in
      let root = widgetRoot(directory), calendar = calendar(root)
      var date = Date(), entries = [Entry(date: date, snapshot: try readSnapshot(directory, at: date))]
      for _ in 0..<32 {
        guard let next = calendar.nextDate(after: date, matching: DateComponents(hour: root["dayStartHour"] as? Int ?? 4, minute: 0), matchingPolicy: .nextTime, repeatedTimePolicy: .first) else { break }
        date = next
        entries.append(Entry(date: date, snapshot: try readSnapshot(directory, at: date)))
      }
      return entries
    }
    completion(Timeline(entries: entries ?? [placeholder(in: context)], policy: .atEnd))
  }
}
private func color(_ light: UInt, _ dark: UInt, scheme: ColorScheme) -> Color {
  let raw = scheme == .dark ? dark : light
  return Color(red: Double((raw >> 16) & 255) / 255, green: Double((raw >> 8) & 255) / 255, blue: Double(raw & 255) / 255)
}
private struct OwndayView: View {
  @Environment(\.colorScheme) private var scheme
  @Environment(\.widgetFamily) private var family
  let entry: Entry
  var body: some View {
    VStack(alignment: .leading, spacing: 8) {
      Text(entry.snapshot.ru ? "Твой день" : "Your day").font(.headline).foregroundStyle(color(__LIGHT_INK__, __DARK_INK__, scheme: scheme))
      if entry.snapshot.habits.isEmpty { Text(entry.snapshot.ru ? "Открой Ownday" : "Open Ownday").font(.caption) }
      ForEach(entry.snapshot.habits.prefix(family == .systemSmall ? 3 : 4)) { habit in
        Button(intent: ToggleHabitIntent(habitId: habit.id, profileId: entry.snapshot.profile, localDate: entry.snapshot.localDate)) {
          HStack(spacing: 8) {
            Image(systemName: habit.done ? "checkmark.circle.fill" : "circle").foregroundStyle(habit.done ? color(__LIGHT_DONE__, __DARK_DONE__, scheme: scheme) : color(__LIGHT_NEUTRAL__, __DARK_NEUTRAL__, scheme: scheme))
            Text(habit.title).font(.caption).lineLimit(1)
            Spacer(minLength: 0)
            if let target = habit.target { Text("\(habit.value.formatted())/\(target.formatted())").font(.caption.monospacedDigit()) }
          }.frame(minHeight: 28).foregroundStyle(color(__LIGHT_INK__, __DARK_INK__, scheme: scheme))
        }.buttonStyle(.plain).accessibilityLabel((habit.done ? (entry.snapshot.ru ? "Отменить: " : "Undo: ") : (entry.snapshot.ru ? "Выполнить: " : "Complete: ")) + habit.title)
      }
    }
    .widgetURL(URL(string: "ownday://"))
    .containerBackground(color(__LIGHT_SURFACE__, __DARK_SURFACE__, scheme: scheme), for: .widget)
  }
}
@main struct OwndayWidget: Widget {
  var body: some WidgetConfiguration {
    StaticConfiguration(kind: "OwndayWidget", provider: Provider()) { OwndayView(entry: $0) }
      .configurationDisplayName("Ownday").description("Your daily habits / Привычки на сегодня")
      .supportedFamilies([.systemSmall, .systemMedium])
  }
}

File: apps/mobile/plugins/OwndayWidget.kt
package __PACKAGE__

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.glance.GlanceId
import androidx.glance.GlanceModifier
import androidx.glance.action.ActionParameters
import androidx.glance.action.actionParametersOf
import androidx.glance.action.clickable
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.GlanceAppWidgetReceiver
import androidx.glance.appwidget.action.ActionCallback
import androidx.glance.appwidget.action.actionRunCallback
import androidx.glance.appwidget.provideContent
import androidx.glance.background
import androidx.glance.layout.*
import androidx.glance.text.FontWeight
import androidx.glance.text.Text
import androidx.glance.text.TextStyle
import androidx.glance.color.ColorProvider
import org.json.JSONArray
import org.json.JSONObject
import java.time.ZoneId
import java.time.ZonedDateTime

internal object WidgetDataLock
internal const val SNAPSHOT_KEY = "ownday.widget.snapshot.v2"
internal const val PENDING_KEY = "ownday.widget.pending.v2"
internal const val PREFS = "ownday_widget"
private val HabitIdKey = ActionParameters.Key<String>("habitId")
private val ProfileKey = ActionParameters.Key<String>("profileId")
private val DateKey = ActionParameters.Key<String>("localDate")
private data class Habit(val id: String, val title: String, var done: Boolean, var value: Double, val target: Double?, var revision: Int, var dependsOn: String?)
private data class Snapshot(val profile: String, val date: String, val ru: Boolean, val habits: List<Habit>)
private fun snapshot(context: Context): Snapshot {
  val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
  return runCatching {
    val root = JSONObject(prefs.getString(SNAPSHOT_KEY, null) ?: return Snapshot("", "", true, emptyList()))
    val now = ZonedDateTime.now(ZoneId.of(root.getString("timezone")))
    val date = (if (now.hour < root.getInt("dayStartHour")) now.toLocalDate().minusDays(1) else now.toLocalDate()).toString()
    val days = root.getJSONArray("days")
    val day = (0 until days.length()).map { days.getJSONObject(it) }.find { it.getString("localDate") == date }
    val values = day?.getJSONArray("habits") ?: JSONArray()
    val habits = (0 until values.length()).map { index ->
      val item = values.getJSONObject(index)
      Habit(item.getString("id"), item.getString("title"), item.optBoolean("done"), item.getDouble("value"), if(item.isNull("target")) null else item.getDouble("target"), item.getInt("revision"), if(item.isNull("dependsOn")) null else item.getString("dependsOn"))
    }
    val pending = JSONArray(prefs.getString(PENDING_KEY, "[]"))
    for (index in 0 until pending.length()) {
      val queued = pending.getJSONObject(index)
      val change = queued.getJSONObject("change")
      if (queued.getString("profileId") != root.getString("profileId") || change.getString("localDate") != date) continue
      val habit = habits.find { it.id == change.getString("habitId") } ?: continue
      val action = change.getJSONObject("action")
      habit.done = action.getString("kind") != "clear"
      habit.value = if (habit.done) action.optDouble("value", habit.target ?: 1.0) else 0.0
      habit.revision += 1
      habit.dependsOn = change.getString("operationId")
    }
    Snapshot(root.getString("profileId"), date, root.getString("locale") == "ru", habits)
  }.getOrDefault(Snapshot("", "", true, emptyList()))
}
private object Palette {
  val surface = ColorProvider(Color(android.graphics.Color.parseColor("__LIGHT_SURFACE__")), Color(android.graphics.Color.parseColor("__DARK_SURFACE__")))
  val ink = ColorProvider(Color(android.graphics.Color.parseColor("__LIGHT_INK__")), Color(android.graphics.Color.parseColor("__DARK_INK__")))
  val neutral = ColorProvider(Color(android.graphics.Color.parseColor("__LIGHT_NEUTRAL__")), Color(android.graphics.Color.parseColor("__DARK_NEUTRAL__")))
  val done = ColorProvider(Color(android.graphics.Color.parseColor("__LIGHT_DONE__")), Color(android.graphics.Color.parseColor("__DARK_DONE__")))
}
private fun number(value: Double) = java.math.BigDecimal.valueOf(value).stripTrailingZeros().toPlainString()
class OwndayWidget : GlanceAppWidget() {
  override suspend fun provideGlance(context: Context, id: GlanceId) {
    val data = synchronized(WidgetDataLock) { snapshot(context) }
    provideContent { WidgetContent(data) }
  }
  @Composable private fun WidgetContent(data: Snapshot) {
    Column(GlanceModifier.fillMaxSize().background(Palette.surface).padding(16.dp)) {
      Text(if(data.ru) "Твой день" else "Your day", style = TextStyle(color = Palette.ink, fontWeight = FontWeight.Bold))
      Spacer(GlanceModifier.height(8.dp))
      if(data.habits.isEmpty()) Text(if(data.ru) "Открой Ownday" else "Open Ownday", style = TextStyle(color = Palette.neutral))
      data.habits.take(4).forEach { habit ->
        Row(GlanceModifier.fillMaxWidth().height(44.dp).clickable(actionRunCallback<ToggleHabitAction>(actionParametersOf(HabitIdKey to habit.id, ProfileKey to data.profile, DateKey to data.date))), verticalAlignment = Alignment.Vertical.CenterVertically) {
          Text(if(habit.done) "✓" else "○", style = TextStyle(color = if(habit.done) Palette.done else Palette.neutral))
          Spacer(GlanceModifier.width(8.dp))
          Text(habit.title, modifier = GlanceModifier.defaultWeight(), style = TextStyle(color = if(habit.done) Palette.neutral else Palette.ink), maxLines = 1)
          if(habit.target != null) Text(" " + number(habit.value) + "/" + number(habit.target), style = TextStyle(color = Palette.neutral))
        }
      }
    }
  }
}
class ToggleHabitAction : ActionCallback {
  override suspend fun onAction(context: Context, glanceId: GlanceId, parameters: ActionParameters) {
    synchronized(WidgetDataLock) {
      val data = snapshot(context)
      // A stale launcher surface must never act on a new account or calendar date.
      if (parameters[ProfileKey] == data.profile && parameters[DateKey] == data.date) {
        val habit = data.habits.find { it.id == parameters[HabitIdKey] }
        if (habit != null) {
          val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
          val pending = JSONArray(prefs.getString(PENDING_KEY, "[]"))
          val action = if (habit.done) JSONObject().put("kind", "clear") else JSONObject().put("kind", "set").put("status", "done").put("value", habit.target ?: 1.0)
          val change = JSONObject().put("operationId", java.util.UUID.randomUUID().toString()).put("habitId", habit.id).put("localDate", data.date).put("baseRevision", habit.revision).put("action", action)
          if (habit.dependsOn != null) change.put("dependsOn", habit.dependsOn)
          pending.put(JSONObject().put("profileId", data.profile).put("change", change).put("createdAt", java.time.Instant.now().toString()))
          check(prefs.edit().putString(PENDING_KEY, pending.toString()).commit()) { "Could not persist widget action" }
        }
      }
    }
    OwndayWidget().update(context, glanceId)
  }
}
class OwndayWidgetReceiver : GlanceAppWidgetReceiver() {
  override val glanceAppWidget = OwndayWidget()
}
